--- elogind-246.9.1/src/login/pam_elogind.c.orig	2020-12-24 11:41:51.396914962 +0100
+++ elogind-246.9.1/src/login/pam_elogind.c	2020-12-24 11:42:14.379396302 +0100
@@ -699,8 +699,8 @@
                 return r;
 
         /* Make most of this a NOP on non-logind systems */
-        if (!logind_running())
-                goto success;
+//        if (!logind_running())
+//                goto success;
 
         /* Make sure we don't enter a loop by talking to
          * elogind when it is actually waiting for the
