--- rustc-1.30.1-src/src/librustc_llvm/lib.rs.old	2018-11-25 12:42:31.851386719 +0100
+++ rustc-1.30.1-src/src/librustc_llvm/lib.rs	2018-11-25 12:56:40.716467824 +0100
@@ -20,6 +20,9 @@
 #[allow(unused_extern_crates)]
 extern crate rustc_cratesio_shim;
 
+// Workaround for rust-lang/rust#39880
+#[link(name = "ffi")] extern {}
+
 // NOTE: This crate only exists to allow linking on mingw targets.
 
 /// Initialize targets enabled by the build script via `cfg(llvm_component = "...")`.
